File: backend/test/battle-rules/expanded-status.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { createBattleState, withStatusApplied, hasStatus, hpOf, apOf } from '../../src/battle-rules/battle-state.js';
import { defaultCombinationBook } from '../../src/battle-rules/combination-book.js';
import { playTurn } from '../../src/battle-rules/turn-engine.js';
import type { BattleState } from '../../src/battle-rules/types.js';

const start = () => createBattleState({playerAId: 'a', playerBId: 'b', currentTurnId: 'a', ap: {a: {max: 5, current: 3}}});
const status = (state: BattleState, target: string, effectId: string, turnsRemaining = 1, damagePerTick = 0) =>
  withStatusApplied(state, target, {effectId, turnsRemaining, damagePerTick});
const hit = (state: BattleState, elementIds: string[]) =>
  playTurn(state, {actorId: state.currentTurnId, elementIds}, defaultCombinationBook).state;

test('silence rejects combos atomically, allows basics and defense', () => {
  const state = status(start(), 'a', 'silence');
  assert.throws(() => hit(state, ['fire', 'wind']), /Silêncio/);
  assert.equal(apOf(state, 'a').current, 3);
  assert.equal(hasStatus(hit(state, ['fire']), 'a', 'silence'), false);
  assert.equal(playTurn(state, {actorId:'a', elementIds:[], kind:'defend'}, defaultCombinationBook).state.currentTurnId, 'b');
});
test('slow removes regeneration; shock increases combo cost only', () => {
  assert.equal(apOf(hit(status(start(), 'a', 'slow'), ['fire']), 'a').current, 3);
  const state = status(start(), 'a', 'shock');
  assert.equal(apOf(hit(state, ['fire', 'wind']), 'a').current, 0);
  assert.equal(apOf(hit(state, ['fire']), 'a').current, 4);
  assert.throws(() => hit(state, ['fire','earth','wind']));
});
for (const [effect, damage] of [['buff', 7], ['debuff', 4]] as const) {
  test(`${effect} modifies direct damage and expires`, () => {
    const state = status(start(), 'a', effect);
    const result = hit(state, ['fire']);
    assert.equal(hpOf(state, 'b').current - hpOf(result, 'b').current, damage);
    assert.equal(hasStatus(result, 'a', effect), false);
  });
}
test('modifiers cancel; wet requires electric hit and respects shield', () => {
  const state = status(status(start(), 'a', 'buff'), 'a', 'debuff');
  assert.equal(hpOf(state, 'b').current - hpOf(hit(state, ['fire']), 'b').current, 5);
  const wet = status(start(), 'b', 'wet', 2);
  const result = hit(wet, ['lightning']);
  assert.equal(hpOf(wet, 'b').current - hpOf(result, 'b').current, 7);
  assert.equal(hasStatus(result, 'b', 'wet'), false);
  assert.equal(hasStatus(hit(wet, ['lightning','shadow']), 'b', 'wet'), true);
  const blocked = hit(status(wet, 'b', 'shield', 2), ['lightning']);
  assert.equal(hpOf(blocked, 'b').current, hpOf(wet, 'b').current);
  assert.equal(hasStatus(blocked, 'b', 'wet'), true);
});
test('poison ticks 2,3,4 then expires', () => {
  let state = status(start(), 'b', 'poison', 3, 2);
  for (const damage of [2,3,4]) {
    const before = hpOf(state, 'b').current;
    state = playTurn(state, {actorId:state.currentTurnId, elementIds:[], kind:'defend'}, defaultCombinationBook).state;
    assert.equal(before - hpOf(state, 'b').current, damage);
  }
  assert.equal(hasStatus(state, 'b', 'poison'), false);
});
test('catalog contains 20 unique recipes executable in reverse order', () => {
  const elements = ['fire','water','wind','ice','nature','lightning','earth','shadow','light','poison'];
  const found = new Set<string>();
  for (let i = 0; i < elements.length; i++) {
    for (let j = i + 1; j < elements.length; j++) {
      for (let k = j; k < elements.length; k++) {
        const ids = k === j ? [elements[i]!, elements[j]!] : [elements[i]!, elements[j]!, elements[k]!];
        const combo = defaultCombinationBook.resolve(ids);
        if (!combo) continue;
        assert.equal(found.has(combo.id), false);
        found.add(combo.id);
        const state = {...start(), ap: {a:{max:5,current:5}, b:{max:5,current:5}}};
        assert.equal(playTurn(state, {actorId:'a', elementIds:ids.reverse()}, defaultCombinationBook).triggeredCombinationId, combo.id);
      }
    }
  }
  assert.equal(found.size, 20);
});

File: packages/battle_engine/test/expanded_status_test.dart
import 'package:battle_engine/battle_engine.dart';
import 'package:test/test.dart';

void main() {
  const a = Combatant(id: 'a', name: 'A');
  const b = Combatant(id: 'b', name: 'B');
  final engine = TurnEngine(defaultCombinationBook);
  BattleState start() => BattleState.start(
    playerA: a,
    playerB: b,
    ap: {a: const ApPool(max: 5, current: 3)},
  );
  BattleState status(
    BattleState state,
    Combatant target,
    StatusEffect effect, {
    int turns = 1,
    int tick = 0,
  }) => state.withStatusApplied(
    target,
    ActiveStatus(effect: effect, turnsRemaining: turns, damagePerTick: tick),
  );
  BattleState hit(BattleState state, List<Element> elements) => engine
      .playTurn(state, TurnAction(actor: state.currentTurn, elements: elements))
      .state;

  test('silence rejects combos atomically but permits basic attacks', () {
    final state = status(start(), a, StatusEffects.silence);
    expect(() => hit(state, [Elements.fire, Elements.wind]), throwsStateError);
    expect(state.apOf(a).current, 3);
    expect(
      hit(state, [Elements.fire]).hasStatus(a, StatusEffects.silence),
      false,
    );
    expect(
      engine.playTurn(state, TurnAction.defend(actor: a)).state.currentTurn,
      b,
    );
  });
  test('slow prevents regeneration; shock adds cost only to combos', () {
    expect(
      hit(status(start(), a, StatusEffects.slow), [
        Elements.fire,
      ]).apOf(a).current,
      3,
    );
    final shocked = status(start(), a, StatusEffects.shock);
    expect(hit(shocked, [Elements.fire, Elements.wind]).apOf(a).current, 0);
    expect(hit(shocked, [Elements.fire]).apOf(a).current, 4);
    expect(
      () => hit(shocked, [Elements.fire, Elements.earth, Elements.wind]),
      throwsStateError,
    );
  });
  for (final entry in [(StatusEffects.buff, 7), (StatusEffects.debuff, 4)]) {
    test('${entry.$1.id} modifies direct damage and expires', () {
      final state = status(start(), a, entry.$1);
      final result = hit(state, [Elements.fire]);
      expect(state.hpOf(b).current - result.hpOf(b).current, entry.$2);
      expect(result.hasStatus(a, entry.$1), false);
    });
  }
  test('buff and debuff cancel; wet is consumed only by an electric hit', () {
    final state = status(
      status(start(), a, StatusEffects.buff),
      a,
      StatusEffects.debuff,
    );
    expect(
      state.hpOf(b).current - hit(state, [Elements.fire]).hpOf(b).current,
      5,
    );
    final wet = status(start(), b, StatusEffects.wet, turns: 2);
    final result = hit(wet, [Elements.lightning]);
    expect(wet.hpOf(b).current - result.hpOf(b).current, 7);
    expect(result.hasStatus(b, StatusEffects.wet), false);
    expect(
      hit(wet, [
        Elements.lightning,
        Elements.shadow,
      ]).hasStatus(b, StatusEffects.wet),
      true,
    );
    final shielded = status(wet, b, StatusEffects.shield, turns: 2);
    expect(
      hit(shielded, [Elements.lightning]).hpOf(b).current,
      wet.hpOf(b).current,
    );
    expect(
      hit(shielded, [Elements.lightning]).hasStatus(b, StatusEffects.wet),
      true,
    );
  });
  test('poison ticks 2, 3, 4 on actions then expires', () {
    var state = status(start(), b, StatusEffects.poison, turns: 3, tick: 2);
    for (final damage in [2, 3, 4]) {
      final before = state.hpOf(b).current;
      state = engine
          .playTurn(state, TurnAction.defend(actor: state.currentTurn))
          .state;
      expect(before - state.hpOf(b).current, damage);
    }
    expect(state.hasStatus(b, StatusEffects.poison), false);
  });
  test('20 distinct recipes resolve regardless of order and execute', () {
    final combos = defaultCombinationBook.combinations;
    expect(combos, hasLength(20));
    expect(combos.map((c) => c.resultId).toSet(), hasLength(20));
    final recipes = <String>{};
    for (final combo in combos) {
      final ids = combo.elements.map((e) => e.id).toList()..sort();
      expect(recipes.add(ids.join('+')), true);
      expect(
        defaultCombinationBook
            .resolve(combo.elements.toList().reversed)
            ?.resultId,
        combo.resultId,
      );
      final state = start().copyWith(
        ap: {
          a: const ApPool(max: 5, current: 5),
          b: const ApPool(max: 5, current: 5),
        },
      );
      expect(
        engine
            .playTurn(state, TurnAction(actor: a, elements: combo.elements))
            .triggeredCombination
            ?.resultId,
        combo.resultId,
      );
    }
  });
}
